osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs

```diff
@@ -112,18 +121,48 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
                 velocityChangeBonus *= Math.Pow(Math.Min(osuCurrObj.StrainTime, osuLastObj.StrainTime) / Math.Max(osuCurrObj.StrainTime, osuLastObj.StrainTime), 2);
             }
 
-            if (osuLastObj.BaseObject is Slider)
+            // Add in acute angle bonus or wide angle bonus + velocity change bonus, whichever is larger.
+            aimStrain += Math.Max(acuteAngleBonus * acute_angle_multiplier, wideAngleBonus * wide_angle_multiplier + velocityChangeBonus * velocity_change_multiplier);
+
+            double sliderJumpBonus = 0;
+            if (osuLastObj.BaseObject is Slider slider)
             {
-                // Reward sliders based on velocity.
-                sliderBonus = osuLastObj.TravelDistance / osuLastObj.TravelTime;
+                // Take just slider heads into account because we're computing sliderjumps, not slideraim
+                sliderJumpBonus = slider_jump_multiplier * aimStrain * sliderJumpsAdjustingMultiplier;
+
+                // Reward more if sliders and circles are alternating (actually it's still lower than several sliders in a row)
+                if (osuLastLastObj.BaseObject is HitCircle)
+                {
+                    double alternatingBonus = slider_jump_multiplier * osuLastObj.JumpDistance / osuLastObj.StrainTime;
+
+                    if (osuLastObj.StrainTime > osuLastLastObj.StrainTime)
+                        alternatingBonus *= Math.Pow(Math.Min(osuCurrObj.StrainTime, osuLastObj.StrainTime) / Math.Max(osuCurrObj.StrainTime, osuLastObj.StrainTime), 2);
+
+                    sliderJumpBonus += alternatingBonus;
+                }
+
+                // If slider was slower than notes before - punish it
+                if (osuCurrObj.StrainTime > osuLastObj.StrainTime)
+                    sliderJumpBonus *= Math.Pow(Math.Min(osuCurrObj.StrainTime, osuLastObj.StrainTime) / Math.Max(osuCurrObj.StrainTime, osuLastObj.StrainTime), 2);
```

Comment:
i'm confused by the comment and formula here. the comment talks about notes before the slider, but the slider is `osuLastObj`. so wouldn't the notes before be `osuLastLastObj`? 

the comment could be less ambiguous too since when i read "slider was slower", i am interpreting this as the slider's velocity. but the code seems to be regarding a slowing down in rhythm

Reply:
this line exist to not buff the cases like "jumps jumps jumps, 1/2 slider, jumps jumps jumps"
now I'm think about it - it's still have 0.25 multiplier, maybe i should drop it to 0